DietPi-Software | Docker-Compose

Joulinar opened the pull request:
**Status**: Testing
- [x] implementation
- [x] testing

**Reference**: https://github.com/MichaIng/DietPi/issues/3078

**Commit list/description**:
- DietPi-Software | Docker-Compose: reserve Software ID 187, get ready for further implementation.
- DietPi-Software | Docker-Compose:  Add dependency for Docker and Python (pip)
- DietPi-Software | Docker-Compose: add install and removal process
- DietPi-Software | Python (pip): enable `pipweels` as additional pip repository for ARMv6 and ARMv7
- DietPi-Software | Docker-Compose: removed dependency `libffi-dev` and `libssl-dev` for aarch64/arm64. It is not required for building docker-compose
- DietPi-Software | Python (pip): move `pipweels` installation on top, to be setup first
- DietPi-Software | Docker-Compose: change software ID from 187 to 134

Commit:
v6.35

DietPi-Software | Docker-Compose: reserve Software ID 187, get ready for further implementation.

Joulinar commented:
@MichaIng 
should we make Docker Compose depending on Docker? Means, to install Docker along with Compose? I mean it doesn't make much sense without 🤣

MichaIng commented:
I would say so, similar to how Portainer does.

Commit:
v6.35

DietPi-Software | Docker-Compose:  Add dependency for Docker and Python (pip)

Commit:
v6.35

DietPi-Software | Docker-Compose: add install and removal process
DietPi-Software | Python (pip): enable pipweels as additional pip repository for ARMv6 and ARMv7

Joulinar commented:
@MichaIng 
for the time being, I added all dependency for `G_HW_ARCH 3 aarch64/arm64`. On my test with RPi4 64bit, not all of them are needed. Unfortunately I don't have other boards to verify with other devices. Still I'm lacking how to simulate other boards on a VM 🤣 

As well I did a first attempt to get `pipweels` activated for ARMv6 and ARMv7

MichaIng commented:
> Unfortunately I don't have other boards to verify with other devices.

I'll test it on other ARMv8 boards, however, RPi4 64-bit should behave the same as it is the same Debian APT repository.

Joulinar commented:
I tested it on RPi4 64bit. There `make` and `gcc` are enough. Not sure about NonRPi boards

nonnorm commented:
> Still I'm lacking how to simulate other boards on a VM

[QEMU](https://qemu.org) is good at that.  It has quite a few ARM and ARM64 emulated boards.  DietPi doesn't have any images for any of them (except for the Raspberry Pis), though it does support the OrangePi PC, which is supported.

Joulinar commented:
I was trying QEMU in past but never succeeded to get it running on my W10 box

Commit:
v6.35

DietPi-Software | Docker-Compose: removed dependency libffi-dev and libssl-dev for aarch64/arm64. It is not required for building docker-compose

Joulinar commented:
during the last couple of days, I did some test installation and for me it was working fine. Would be good to have a 2nd pair of eyes checking. Especially on nonRPi boards.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -6976,7 +7002,18 @@ exec sudo -u $ha_user dash -c '$ha_pyenv_activation; exec pip3 install -U homeas
+				
+					echo '[global]
+extra-index-url=https://www.piwheels.org/simple
+' > /etc/pip.conf
```
Let's do that step prior to `get-pip.py`, as, if I'm not wrong, `get-pip.py` makes use of pre-compiled wheels as well. At least it does not hurt.

Joulinar replied:
ok changes, looks good

```
[  OK  ] DietPi-Software | curl -sSfL https://bootstrap.pypa.io/get-pip.py -o get-pip.py
[ INFO ] DietPi-Software | python3 ./get-pip.py, please wait...
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting pip
  Downloading https://www.piwheels.org/simple/pip/pip-21.0-py3-none-any.whl (1.5 MB)
Collecting setuptools
  Downloading https://www.piwheels.org/simple/setuptools/setuptools-52.0.0-py3-none-any.whl (784 kB)
Collecting wheel
  Downloading https://www.piwheels.org/simple/wheel/wheel-0.36.2-py2.py3-none-any.whl (35 kB)
Installing collected packages: wheel, setuptools, pip
Successfully installed pip-21.0 setuptools-52.0.0 wheel-0.36.2
[  OK  ] DietPi-Software | python3 ./get-pip.py
[  OK  ] DietPi-Software | rm get-pip.py
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1347,6 +1347,16 @@ DietPi-Software will decrypt and use it for software installs. You can change it
 		# - Bullseye
 		aSOFTWARE_AVAIL_G_DISTRO[$software_id,6]=0
+		#------------------
+		software_id=187
```
ID 134 can be used now, that it became free one version before.

Joulinar replied:
done

Joulinar committed:
v6.35

DietPi-Software | Python (pip): move pipweels installation on top, to be setup first
DietPi-Software | Docker-Compose: change software ID from 187 to 134

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -15257,6 +15295,15 @@ _EOF_
+
+			Banner_Uninstalling
+			
+			G_EXEC_OUTPUT=1 G_EXEC pip3 uninstall -y docker-compose
```
Shall we allow this failing, in case pip got removed already?
```
G_EXEC_NOHALT=1 G_EXEC_OUTPUT=1 G_EXEC pip3 uninstall -y docker-compose
```
It will still print a red failure but will skip the error prompt and continue. I think for uninstalls it's fine when something fails, or the other way round, it is annoying if an uninstall failed because something has already been uninstalled manually or files gone lost or such, what you think?
_______
Alternatively `G_EXEC_NOEXIT=1` to still show the prompt, but the exit button is an ignore button, so the script will continue, which is true as well when it got called non-interactively (no STDIN or with `G_INTERACTIVE=1`).

Joulinar replied:
well hard to say, if we would allow it failing but continue uninstall, we will end up with `docker-compose` still being installed but we would have marked it as uninstalled on `/boot/dietpi/.installed`. Isn't it?

MichaIng replied:
Yes exactly. Currently most uninstalls are not error-handled and I didn't actively add strict error-handling since I'm regularly annoyed by e.g. DEB package uninstall scripts that fail if certain files of the package are already missing. In such case all further APT progress is blocked, so it's much worse than it would be in our case. I remember a nightmare with some python packages, where it was nearly impossible to perform that uninstall since malformed files in some of them broke their uninstall which then blocked the other uninstall with broken dependencies error and so on. And reinstall didn't work as well because of broken dependencies. I think 90% of users would have needed to install Debian from scratch, which is insane when it is only about the ability to remove all package files so that APT/DPkg are able to attempt a reinstall or clean up dependencies. So while it's good to aim a clean uninstall, it should be at least possible to continue if something fails, as worst thing is a few left files that usually do not hurt or could/need to be be dealt with manually anyway.
________
I now went with NOEXIT, so that it's still an error prompt but possible to ignore.

Joulinar replied:
ok than let's go with it this way

MichaIng commented:
Works very well on x86_64, used it to run the Pi-hole Docker container to do some testing there 👍. A quite simple install anyway, thanks to pip. Let me do some minor code changes, otherwise ready to be merged.

Joulinar commented:
> Works very well on x86_64

🤔 well `piwheels` config file is created for ARMv6 and ARMv7 only 😉 
Or do you mean `docker-compose`?

MichaIng committed:
v6.35

+ Minor

MichaIng commented:
Works great on VM and ARMv7. I'll merge.

Joulinar commented:
ok I verified it once more on my RPi3B+ and first I was irritated by the output. Was not expecting the word-wrap to see 😄 

```
 DietPi-Software
─────────────────────────────────────────────────────
 Mode: Installing Python 3 pip: Python 3 package installer

[      ] DietPi-Software | eval echo -e '[global]
[  OK  ] DietPi-Software | eval echo -e '[global]
extra-index-url=https://www.piwheels.org/simple/' > /etc/pip.conf
[  OK  ] DietPi-Software | Checking URL: https://bootstrap.pypa.io/get-pip.py
```

but it is working well 👍

MichaIng commented:
Indeed that looks strange. Also the notification function that aims to replace the execution animation doesn't recognise manual "\n" linebreaks, so the first line is not overprinted as expected.

I just fixed that within G_EXEC: https://github.com/MichaIng/DietPi/commit/13a0edf7b20710bee7d60cf637a3330e79fa4449
When the passed command contains any escape sequence (like \n newline), that sequence is part of the executed command, so should be shown in the output like that and not expanded through G_DIETPI-NOTIFY to an actual newline. That is quite easy by escaping all backslash characters with a second backslash.

Only left glitch: In the error prompt, the command will still be shown with an actual newline, as whiptail does not support escaping them at all. Like in console `echo -e 'line1\\nline2'` can be used to print "\n" literally, but in whiptail this leads to a backslash followed by an actual newline... But I think it's rare and unimportant enough to be ignored 😄.

Joulinar commented:
way better now. well done 😃

```
 DietPi-Software
─────────────────────────────────────────────────────
 Mode: Installing Python 3 pip: Python 3 package installer

[  OK  ] DietPi-Software | eval echo -e '[global]\nextra-index-url=https://www.piwheels.org/simple/' > /etc/pip.conf
[  OK  ] DietPi-Software | Checking URL: https://bootstrap.pypa.io/get-pip.py
```
